File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function(root) {
    let ans = []
    if(root === null) return ans;
    let curr;

    while(root !== null) {
        if(root.left !== null) {
            curr = root.left;
            while(curr.right !== null && curr.right !== root) {
                curr = curr.right;
            }

            if(curr.right != null) {
                curr.right = null;
                root = root.right;
            }
            else {
                ans.push(root.val);
                curr.right = root;
                root = root.left;
            }
        }
        else {
            ans.push(root.val);
            root = root.right;
        }
    }

    return ans;
};

File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        if(root == null) return ans;
        TreeNode curr;

        while(root != null) {
            if(root.left != null) {
                curr = root.left;
                while(curr.right != null && curr.right != root) {
                    curr = curr.right;
                }

                if(curr.right == null) {
                    ans.add(root.val);
                    curr.right = root;
                    root = root.left;
                }
                else {
                    curr.right = null;
                    root = root.right;
                }
            }
            else {
                ans.add(root.val);
                root = root.right;
            }
        }

        return ans;
    }
}